### PR TITLE
[WIP] new backward

### LINKED ISSUE #6600: Problems in current backward implementation
Current backward implementation is faced with several problems.

1. The design of current backward didn't think much about nested blocks. To make it working on `WhileOp` and `IfElesOp`, we have to add many tricks into it, which makes our code extremely hard to read.

2. Incomplete implementation of backward pruning. We are simply using a `set` to record names of variables which do not have gradients. This strategy can not disgust variables in different blocks with the same name and causes bugs in some special cases.

3. In some cases, such as error clipping, we need to insert some other operators(`ClipOp`) between all backward operators. It's not easy to implement these functions in current C++ code.

### PR DESCRIPTION
**Pull request by JiayiFeng**
fixes #6600

### CODE REVIEW

File: python/paddle/v2/fluid/backward.py [hunk 1]
```diff
@@ -1,9 +1,200 @@
...
+def _strip_grad_suffix_(name):
+    pos = name.find(core.grad_var_suffix())
+    return name[:pos] if pos != -1 else name
+
+
+def _append_grad_suffix_(name):
+    return name + core.grad_var_suffix()
+
+
+def _append_backward_ops_(target,
```
Comment (gongweibao) [inline python/paddle/v2/fluid/backward.py]: This function is so long to read.

Comment (JiayiFeng) [reply python/paddle/v2/fluid/backward.py]: Yes. I'm refactoring it.

**Comment by emailweixu** [comment]
Does this mean that the previous append_backward implemented in C++ is no longer needed?
A related question is that if later we decide to have a different language binding, does this logic have to be re-implemented again in a different language?

File: python/paddle/v2/fluid/backward.py [hunk 2]
```diff
@@ -1,17 +1,209 @@
...
+def append_backward(loss, parameter_list=None, no_grad_set=None):
     """
     Create and add gradient Operators in BlockDesc to compute
     gradients of `loss` for parameters in parameter_list
 
     :param loss: an variable generated by cost function.
     :type loss: Variable
-    :param no_grad_set: variable that should not create gradient
-    :type no_grad_set: set
+    :param no_grad_dict: variable that should not create gradient
```
Comment (emailweixu) [inline python/paddle/v2/fluid/backward.py]: At line 225, the no_grad variable name has suffix grad_suffix. Does the name if no_grad_set also need to include grad_suffix?

**Comment by JiayiFeng** [comment]
Yes, the code in C++ is no longer needed. 
Yes, this logic has to be re-implemented if we decide to have another language binding.

The purpose of this refactor is making backward supporting callback functions, which is required by error clip and gradient reduce in distributed training.

**Pull request merged**